internals/daemon/api_notices_test.go

```diff
@@ -107,20 +116,124 @@ func (s *apiSuite) testNoticesFilter(c *C, makeQuery func(after time.Time) url.V
 	})
 }
 
+func (s *apiSuite) TestNoticesFilterMultipleUsers(c *C) {
+	s.daemon(c)
+
+	st := s.d.overlord.State()
+	st.Lock()
+	addNotice(c, st, 0, state.ChangeUpdateNotice, "123", nil)
+	time.Sleep(time.Microsecond)
+	addNotice(c, st, 1000, state.CustomNotice, "a.b/x", nil)
+	time.Sleep(time.Microsecond)
+	addNotice(c, st, -1, state.WarningNotice, "danger", nil)
+	st.Unlock()
+
+	noticesCmd := apiCmd("/v1/notices")
+
+	// Test that root user may see all notices, and may filter on any user.
+	userRemoteAddr := "pid=100;uid=0;socket=;"
+	req, err := http.NewRequest("GET", "/v1/notices", nil)
+	c.Assert(err, IsNil)
+	req.RemoteAddr = userRemoteAddr
+	rsp, ok := noticesCmd.GET(noticesCmd, req, nil).(*resp)
+	c.Assert(ok, Equals, true)
+
+	c.Check(rsp.Type, Equals, ResponseTypeSync)
+	c.Check(rsp.Status, Equals, http.StatusOK)
+	notices, ok := rsp.Result.([]*state.Notice)
+	c.Assert(ok, Equals, true)
+	c.Assert(notices, HasLen, 3)
+	n := noticeToMap(c, notices[0])
+	c.Assert(n["user"], Equals, 0.0)
+	n = noticeToMap(c, notices[1])
+	c.Assert(n["user"], Equals, 1000.0)
+	n = noticeToMap(c, notices[2])
+	c.Assert(n["user"], Equals, -1.0)
+
+	for _, users := range [][]int64{{0, 1000, -1}, {0, 1000}, {0, -1}, {1000, -1}, {0}, {1000}, {-1}} {
+		usersStrings := make([]string, 0, len(users))
+		for _, uid := range users {
+			usersStrings = append(usersStrings, fmt.Sprintf("users=%d", uid))
+		}
+		reqUrl := fmt.Sprintf("/v1/notices?%s", strings.Join(usersStrings, "&"))
+		req, err := http.NewRequest("GET", reqUrl, nil)
+		c.Assert(err, IsNil)
+		req.RemoteAddr = userRemoteAddr
+		rsp, ok := noticesCmd.GET(noticesCmd, req, nil).(*resp)
+		c.Assert(ok, Equals, true)
+
+		c.Check(rsp.Type, Equals, ResponseTypeSync)
+		c.Check(rsp.Status, Equals, http.StatusOK)
+		notices, ok := rsp.Result.([]*state.Notice)
+		c.Assert(ok, Equals, true)
+		c.Assert(notices, HasLen, len(users))
+		for i, uid := range users {
+			n := noticeToMap(c, notices[i])
+			c.Assert(n["user"], Equals, float64(uid))
+		}
+	}
+
+	// Test that non-root user by default only sees their notices and notices
```

Comment:
I think it'd be better if each of these "Test ..." blocks should be its own test function, even if there's a bit more repetition.